balancer/weightedroundrobin/balancer.go

```diff
@@ -141,273 +147,311 @@ func (bb) Name() string {
 
-		var wsc *weightedSubConn
-		wsci, ok := b.subConns.Get(addr)
-		if ok {
-			wsc = wsci.(*weightedSubConn)
-		} else {
-			// addr is a new address (not existing in b.subConns).
-			var sc balancer.SubConn
-			sc, err := b.cc.NewSubConn([]resolver.Address{addr}, balancer.NewSubConnOptions{
-				StateListener: func(state balancer.SubConnState) {
-					b.updateSubConnState(sc, state)
-				},
-			})
-			if err != nil {
-				b.logger.Warningf("Failed to create new SubConn for address %v: %v", addr, err)
+	for _, childState := range childStates {
+		if childState.State.ConnectivityState == connectivity.Ready {
+			ewv, ok := b.endpointToWeight.Get(childState.Endpoint)
+			if !ok {
+				// Should never happen, simply continue and ignore this endpoint
+				// for READY pickers.
 				continue
 			}
-			wsc = &weightedSubConn{
-				SubConn:           sc,
-				logger:            b.logger,
-				connectivityState: connectivity.Idle,
-				// Initially, we set load reports to off, because they are not
-				// running upon initial weightedSubConn creation.
-				cfg: &lbConfig{EnableOOBLoadReport: false},
-
-				metricsRecorder: b.metricsRecorder,
-				target:          b.target,
-				locality:        b.locality,
-			}
-			b.subConns.Set(addr, wsc)
-			b.scMap[sc] = wsc
-			b.csEvltr.RecordTransition(connectivity.Shutdown, connectivity.Idle)
-			sc.Connect()
+			ew := ewv.(*endpointWeight)
+			readyPickersWeight = append(readyPickersWeight, pickerWeightedEndpoint{
+				picker:           childState.State.Picker,
+				weightedEndpoint: ew,
+			})
 		}
-		// Update config for existing weightedSubConn or send update for first
-		// time to new one.  Ensures an OOB listener is running if needed
-		// (and stops the existing one if applicable).
-		wsc.updateConfig(b.cfg)
+	}
+	// If no ready pickers are present, simply defer to the round robin picker
+	// from endpoint sharding, which will round robin across the most relevant
+	// pick first children in the highest precedence connectivity state.
+	if len(readyPickersWeight) == 0 {
+		b.ClientConn.UpdateState(balancer.State{
+			ConnectivityState: state.ConnectivityState,
+			Picker:            state.Picker,
+		})
+		return
 	}
 
-	// Loop through existing subconns and remove ones that are not in addrs.
-	for _, addr := range b.subConns.Keys() {
-		if _, ok := addrsSet.Get(addr); ok {
-			// Existing address also in new address list; skip.
-			continue
-		}
-		// addr was removed by resolver.  Remove.
-		wsci, _ := b.subConns.Get(addr)
-		wsc := wsci.(*weightedSubConn)
-		wsc.SubConn.Shutdown()
-		b.subConns.Delete(addr)
+	p := &picker{
+		v:               rand.Uint32(), // start the scheduler at a random point
+		cfg:             b.cfg,
+		weightedPickers: readyPickersWeight,
+		metricsRecorder: b.metricsRecorder,
+		locality:        b.locality,
+		target:          b.target,
 	}
+
+	var ctx context.Context
+	ctx, b.stopPicker = context.WithCancel(context.Background())
```

Comment:
Oh...did I use a context here?  :thinking: I think this should be replace with an `Event` (or a signaling channel).

Reply:
Yeah you did use a context here. If the main reason you want to switch this is goa conformance tests I think this part won't get searched. It graph searches down the exact symbols it calls so NewClient/Server, which is why it hits the xDS Client, but this happens underneath the hood and won't get searched.

So if this is the way you did it it seems like the right way to do it? Give it a context that you can close from here and close all the resources. This relates to Easwar's xDS Client context scoping we might want to refactor with Purneshes changes, what's the best practices for context scoping within a tree and within a tree that then needs to be able to deallocate resources it's spawned.

Let me know if you feel strongly about this and I can change it tomorrow. I got to all other comments.

Reply:
> It graph searches down 

I'm definitely concerned about this part.  If it ever needs to use WRR (it doesn't today AFAIK) then I think it will scan and see this.  But we can fix separately.

Reply:
I can switch this to a grpcsync.Event and receive off that channel instead of ctx.Done() if you want me too.